chain/tipset.go

```diff
@@ -157,6 +163,32 @@ func (c *TipSetCache) Reset() {
 	c.len = 0
 }
 
+// Warm fills the TipSetCache with confidence tipsets so that subsequent calls to Add return a tipset.
+func (c *TipSetCache) Warm(ctx context.Context, head *types.TipSet, fn func(ctx context.Context, tsk types.TipSetKey) (*types.TipSet, error)) error {
+	cur := head
+	tss := make([]*types.TipSet, 0, c.Confidence())
+	for i := 0; i < c.Confidence(); i++ {
+		cur, err := fn(ctx, cur.Parents())
+		if err != nil {
+			return err
+		}
+		tss = append(tss, cur)
+		if cur.Height() == 0 {
```

Comment:
For safety you should probably make this check at the start of the loop. head may be genesis